File: 패스트캠퍼스/코테2/3.py
def solution(N):
    col, diag1, diag2 = set(), set(), set()
    answer = 0
    def dfs(N, y, col, diag1, diag2):
        nonlocal answer
        if y == N:
            answer += 1
            return

        for x in range(N):
            if x in col or (x + y) in diag1 or (x - y) in diag2:
                continue
            col.add(x)
            diag1.add(x + y)
            diag2.add(x - y)
            dfs(N, y + 1, col, diag1, diag2)
            col.remove(x)
            diag1.remove(x + y)
            diag2.remove(x - y)
    dfs(N, 0, col, diag1, diag2)
    return answer

print(solution(12))

File: 패스트캠퍼스/자체 코테 두번쨰/3.js
function solution(height) {
  const answer = [];

  const stack = [];

  for (let i = 0; i < height.length; i++) {
    if (stack.length === 0) {
      stack.push([height[i], i + 1]);
      answer.push(0);
    } else {
      while (stack.length !== 0 && stack[stack.length - 1][0] <= height[i]) {
        stack.pop();
      }
      if (stack.length === 0) {
        stack.push([height[i], i + 1]);
        answer.push(0);
      } else {
        answer.push(stack[stack.length - 1][1]);
        stack.push([height[i], i + 1]);
      }
    }
  }

  return answer;
}

console.log(solution([50, 57, 52, 53, 51]));
